Controlling User Interaction (#228)
camaraproject/IdentityAndConsentManagement: documentation/CAMARA-API-access-and-user-consent.md

Purpose must be specified in the authorization request for a CAMARA Three-Legged Access Token, this is achieved by extending the `scope` parameter as defined in [Purpose as a Scope](CAMARA-Security-Interoperability.md#purpose-as-a-scope) section of the CAMARA Security and Interoperability Profile.

> ## Controlling User-Interaction

Reviewer: I think this section is missed in the Table of Contents.

> In CIBA User-Interaction in initiated by the API Producer if Consent is needed and NOT granted. The `login_hint` parameter is mandatory in CIBA and is used to identify the authentication devicde. Consent is captured out-of-band by e.g. sending a push notification, or by sending an SMS with a code or link, or by sending a message by some other means to the authentication devide. CAMARA does not mandate a communication method or any message format or content.

Reviewer 2: It is not mandatory for the API provider to seek user or end-user consent. They can just reject the authentication request instead.

The CIBA `login_hint` parameter itself will not necessarily be being used by the **API consumer** to identify an authentication device. They may be using it solely to identify a subscription. If the API Provider then uses that parameter to identify and authentication device, that is up to them. But they might alternatively identify an e-mail address, for example.

It is important to understand that, for some APIs, it is user, and not end user, consent that is required, even if the end user is different from the user.

```suggestion
If the API Consumer is using CIBA, User-Interaction MAY be initiated by the API Producer if Consent is needed and has NOT been granted. The `login_hint` parameter is mandatory in CIBA, and MAY be used by the API Producer to identify an authentication device for out-of-band consent capture. Consent MAY be captured out-of-band by, for example, sending a push notification, or by sending an SMS with a code or link, or by sending a message by some other means to the identified authentication device. CAMARA does not mandate any communication method nor any message format nor any message content.
```

Reviewer: Please consider my format suggestions in other comments...

> In OIDC Authorization Code flow User-Interaction is initiated by the API Producer if consent is needed. The Authorization Server obtains End-User Consent/Authorization accorinding to [OIDC section 3.2.1.4](https://openid.net/specs/openid-connect-core-1_0.html#Consent).

Reviewer: It would be advisable to maintain consistency with the rest of the document by using the same terms throughout. With regard to the term 'API producer', it is my understanding that when used in this context, the document states 'API provider' rather than 'API producer'.

Author: I searched the current document for "API pr" and found 4 hits, two "API provider", and two "API producer". I am going to change all hits to "API provider". Commonalities has "API Provider" in its glossary https://github.com/camaraproject/Commonalities/blob/main/documentation/Glossary.md
So, "API Provider" it is, right?! Should I change to C also in API consumer (4 occurrences)?

Author: I changed API producer to API Provider in the new section.
Should I change API producer everywhere? There is also API invoker, which should probably be API Consumer, right?

Author: Because of the Commonalities glossary I now changed API Invoker to API Consumer everywhere. Hope that is OK with everybody.
Also changed API [p|P]roducer to API Provider.
Also changed API [c|C]onsumer to API Consumer.

Reviewer: I see you changed `API invoker` to `API Consumer`, but I'm afraid that if you look for just `Invoker`, there are 16 more coincidences in the document. I'd suggest replacing `Invoker` with `API Consumer`, and `invoker_callback` with `api_consumer_callback`.

Author: I changed Invoker to API consumer and invoker_callback to api_consumer_callback

> In OIDC Authorization Code flow User-Interaction is initiated by the API Producer if consent is needed. The Authorization Server obtains End-User Consent/Authorization accorinding to [OIDC section 3.2.1.4](https://openid.net/specs/openid-connect-core-1_0.html#Consent).

Reviewer: ```suggestion
In OIDC Authorization Code flow User-Interaction is initiated by the API provider if consent is needed. The Authorization Server obtains End-User Consent/Authorization according to [OIDC section 3.2.1.4](https://openid.net/specs/openid-connect-core-1_0.html#Consent).
```

> If the API Consumer wants to make sure that no User-Interaction happens they MUST use the `prompt` parameter with a value of `none`. Authentication Error Response values (`interaction_required`, `login_required`, `account_selection_required`, `consent_required`) for `prompt=none` are defined in [OIDC section 3.1.2.6](https://openid.net/specs/openid-connect-core-1_0.html#AuthError).

Reviewer: ```suggestion
If the API consumer wants to make sure that no User-Interaction happens they MUST use the `prompt` parameter with a value of `none`. Authentication error response values (`interaction_required`, `login_required`, `account_selection_required`, `consent_required`) for `prompt=none` are defined in [OIDC section 3.1.2.6](https://openid.net/specs/openid-connect-core-1_0.html#AuthError).
```

Reviewer: Capital letters are for terms defined in "Terms and Definitions" section.

Author: API Consumer is in the Commonalities glossary. I'll change the "error response". I copied Authentication Error Response from [OIDC Authentication Error Response](https://openid.net/specs/openid-connect-core-1_0.html#AuthError) were it is written as "Authentication Error Response" not only in the headline but in the text too. But I do not care about this too much.

> It is up to the operator to make a decision when consent collection is needed or not, based on the scope(s)/purpose declared by the API Consumer and aligned with local legislation, ensuring that all operators under the same regulatory framework adopt a **consistent** approach.

Reviewer: ```suggestion
It is up to the operator to make a decision when consent collection is needed or not, based on the scope(s)/purpose declared by the API consumer and aligned with local legislation, ensuring that all operators under the same regulatory framework adopt a **consistent** approach.
```

> The API Consumer can use the `prompt` parameter value `consent` to ask the API producer for End-User Consent collection.

Reviewer: ```suggestion
The API consumer can use the `prompt` parameter value `consent` to ask the API provider for End-User Consent collection.
```

If the API Consumer wants to make sure that no User-Interaction happens they MUST use the `prompt` parameter with a value of `none`. Authentication error response values (`interaction_required`, `login_required`, `account_selection_required`, `consent_required`) for `prompt=none` are defined in [OIDC section 3.1.2.6](https://openid.net/specs/openid-connect-core-1_0.html#AuthError).

It is up to the operator to make a decision when consent collection is needed or not, based on the scope(s)/purpose declared by the API Consumer and aligned with local legislation, ensuring that all operators under the same regulatory framework adopt a **consistent** approach.

The API Consumer can use the `prompt` parameter value `consent` to ask the API Provider for End-User Consent collection.
> The Authorization Server SHOULD prompt the End-User for consent before returning information to the Client. If it cannot obtain consent, it MUST return an error, typically consent_required.

The `prompt` parameter is defined in [OIDC section 3.1.2.1](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest).

> OIDC specifies that the `prompt` parameter is [mandatory to implement for all OpenId Providers](https://openid.net/specs/openid-connect-core-1_0.html#ServerMTI).

Reviewer 2: This sentence might be misleading for API consumers, as an API provider can "implement" support for `prompt=consent` simply by always returning an error and not seeking End User consent at all.  Only `prompt=none` has any defined behaviour that the API provider MUST implement.

I would remove this sentence.
